### PR TITLE
Feature/csr retrieve entity

### PR DESCRIPTION
**Pull request by thomasBousselin**
Context Source registration working for retrieve entity.

some minimal testing ok with postman and the test suite

missing : 
- local parameter
- unit test
- more complete testing with the test suite

### CODE REVIEW

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:111]: please don't throw exceptions 🙏

**Review by bobeal — COMMENTED** [review]
A first quick review

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/model/CSRFilters.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/model/CSRFilters.kt]: **consistency** with other services. And CSRService is only 230 lines...

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/model/CSRFilters.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/model/CSRFilters.kt:5]: - it is a TODO that you plan to resolve in the current PR: OK
- it is a TODO that is not planned to be resolved in the current PR: open an issue or add a note in the specification doc
- it is a comment about some specific piece of code: it is not a TODO then

tracking issues is already complex, don't need to transform the code into yet another issue tracker

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt:187]: at the end, I'm not sure that it will be possible to have the same code for both. they serve different purposes and have different inputs...

**Review by bobeal — COMMENTED** [review]
PS: I still have to look at the merge algorithm

File: search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:204]: OK, it is managed when merging the entity

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt]: yes, 404 is not considered as an error (but it says that CSR **may** hold the data). and it seems like errors should be swallowed for retrieve entity (and consumption operations in general).

better handle the `timesSent` / `lastSucces` / `lastFailure` / ... in this PR.

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:138]: not supposed to happen does not mean it will never happen. and throwing an exception will stop the request and the end user will receive a 500. not the expected behavior?

File: search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtilsTests.kt
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtilsTests.kt:25]: `TemporalPaginationService`?

Comment (bobeal) [reply search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtilsTests.kt:25]: Actually, you don't even need a Spring Boot test here, a simple JUnit test should be enough?

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt:249]: `ngsiLdDateTime()`

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt:253]: `ngsiLdDateTime()`

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:166]: It can't happen as long as the function is call for non CORE attribute. (which is his purpose) 
If the error happen that means its a development issue.
I could replace it with : 
`        is Map<*, *> -> {
            attribute as CompactedAttributeInstance
            mapOf(attribute[NGSILD_DATASET_ID_TERM] as? String to attribute)
        }
        else -> {
            attribute as CompactedAttributeInstances
            attribute.associateBy { it[NGSILD_DATASET_ID_TERM] as? String }
        }`
        to avoid calling explicitely the throw if you prefer. (but it will have the same behavior)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:43]: the correct return type should be `Either<NGSILDWarning, CompactedEntity?>` (as "responding with no data and the HTTP status code 404 - Not Found should not be considered as abnormal behaviour for distributed operations")

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:119]: `type1`, `type2` is harder to understand, something like `current`, `new`? also the function is used to merge types and scopes, so parameters can't be named `typeX`

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:166]: Or it is a context source returning an invalid payload (and should be translated to a 111 - Revalidation Failed)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:133]: same here for the naming

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:147]: that should at least be commented

File: shared/src/main/kotlin/com/egm/stellio/shared/model/ApiExceptions.kt
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/model/ApiExceptions.kt:9]: let's discuss about this, don't really like seeing the warnings "polluting" the whole error hierarchy (whereas it is specific to distOps and HTTP binding)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:118]: note that this will not work if entity is asked for using the simplified representation (and what about the GeoJSON representation?). or we always ask for the normalized representation to context sources and the final representation is done in the main context broker (as is currently done in `EntityHandler`)? (in this case, we should take care to not transmit the request parameter to the context source).

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:41]: this deserves a test to be sure we get the expected result if one or two merges fail

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:50]: when fixing merge conflicts, the terminology is `current` and `incoming`

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:78]: since both `List` and `Set` inherit from `Iterable`, you could work with this type and avoid the `List` <-> `Set` transformations and make this a bit more digest ;)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:118]: If we query the CSR for the simplified representation (or geojson) we don't get the observedAt information so we can't merge the data properly. So i think the final representation shall be in the main broker. ( and i will remove the request param from the CSRrequest)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:79]: I could replace it by :
```
 currentValue == remoteValue -> currentValue
        currentValue is Iterable<*> && remoteValue is List<*> -> (currentValue.toSet() + remoteValue.toSet())
        currentValue is Iterable<*> -> (currentValue.toSet() + remoteValue)
        remoteValue is Iterable<*> -> (remoteValue.toSet() + currentValue)
        else -> listOf(currentValue, remoteValue)
```
But at the end it is no longer a list and it could impact the rest of the code.

Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:79]: as you know, not really fan of this function but not a big deal. let's focus on more important things.

File: search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceCallerTests.kt
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceCallerTests.kt:149]: the value of the header is `application/geo+json` (and not `GEO_JSON_MEDIA_TYPE`)

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:44]: if it is just to be able to call it in the test, you can instead make it `internal`

File: search-service/src/test/kotlin/com/egm/stellio/search/entity/web/EntityHandlerTests.kt
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/entity/web/EntityHandlerTests.kt:812]: ... the warning sent by ...

File: search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:207]: you don't initialize `csf`, not sure something will happen in the service?

Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:207]: it should be initialized with `csf=(operations==federationOps|operations==retrieveEntity`)` (even if when querying you only deal with `operations`)

Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:207]: I choose to have both csf for getCSRs endpoint and operations for inside calls. Will you prefer to only have csf.

Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:207]: cleaner to have only `csf`, no ? (calling from inside as if you were calling from outside)

Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/entity/web/EntityHandler.kt:207]: i thought and then i type "operations==federationOps|operations==retrieveEntity" and i changed my mind. you have no autocompletion and can't be sure your csf is valid.
A solution is to make a utils function csfBuilder that take a list of operation and build it for you.

File: search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtilsTests.kt
Comment (bobeal) [inline search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtilsTests.kt:66]:
```suggestion
    private val invalidEntityWithLastName = entityWithLastName.toMutableMap()
```

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt
Comment (bobeal) [inline search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationService.kt:233]: add at least one test for this `csf` thing (don't think I have seen one)

**Review by ranim-n — APPROVED** [review]

File: search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt
Comment (thomasBousselin) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:101]: Dans detekt il y a la regle OptionnalUnit qui bloque { Unit }.
Est ce que je désactive la règle? Ou alors on change la regle Sonar (au choix).

Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt:101]: `Unit` (and not `{ Unit }`) seems to work

File: search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceCallerTests.kt
Comment (thomasBousselin) [reply search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceCallerTests.kt:158]: Yes, because we can't merge different GEOJSON payload. moreover the GEOJSON is treated in the toFinalRepresentation which is after all the entity are merged.

File: search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationServiceTests.kt
Comment (thomasBousselin) [reply search-service/src/test/kotlin/com/egm/stellio/search/csr/service/ContextSourceRegistrationServiceTests.kt]: I renamed it `query CSR on entities ids` is it better?

**Review by bobeal — APPROVED** [review]

**Pull request merged**
